### ISSUE TITLE
Trying to understand yolo_loss

### ISSUE THREAD

**Issue by ethanyanjiali** [zzh8829/yolov3-tf2#115]
first, thanks for answering my last question.

In the following code, initially, `y_pred` is just the prediction from the model and `y_true` contains image coordinates of ground truth box (relative to the whole image)

however, you applied `yolo_boxes` on `y_pred` and changed it to image coordinates `(bx, by, bw, bh)`, and then transformed `y_true` from image coordinates to cell offset `(tx, ty, tw, th)`

later in the code, the loss is calculated upon `y_pred` in format of `(bx, by, bw, bh)` and `y_true` in format of `(tx, ty, tw, th)`.  shouldn't they all be in one format? from my understanding, the training should use `(tx, ty, tw, th)` as it's easier to converge. only the inference uses `(bx, by, bw, bh)`

```python
    def yolo_loss(y_true, y_pred):
        # 1. transform all pred outputs
        # y_pred: (batch_size, grid, grid, anchors, (x, y, w, h, obj, ...cls))
        pred_box, pred_obj, pred_class, pred_xywh = yolo_boxes(
            y_pred, anchors, classes)
        pred_xy = pred_xywh[..., 0:2]
        pred_wh = pred_xywh[..., 2:4]

        # 2. transform all true outputs
        # y_true: (batch_size, grid, grid, anchors, (x1, y1, x2, y2, obj, cls))
        true_box, true_obj, true_class_idx = tf.split(
            y_true, (4, 1, 1), axis=-1)
        true_xy = (true_box[..., 0:2] + true_box[..., 2:4]) / 2
        true_wh = true_box[..., 2:4] - true_box[..., 0:2]

        # give higher weights to small boxes
        box_loss_scale = 2 - true_wh[..., 0] * true_wh[..., 1]

        # 3. inverting the pred box equations
        grid_size = tf.shape(y_true)[1]
        grid = tf.meshgrid(tf.range(grid_size), tf.range(grid_size))
        grid = tf.expand_dims(tf.stack(grid, axis=-1), axis=2)
        true_xy = true_xy * tf.cast(grid_size, tf.float32) - \
            tf.cast(grid, tf.float32)
        true_wh = tf.math.log(true_wh / anchors)
        true_wh = tf.where(tf.math.is_inf(true_wh),
                           tf.zeros_like(true_wh), true_wh)
```

**Comment by ethanyanjiali** [comment]
to clarify my question further, i'm using the original yolo code as an example:
as we can see, `get_yolo_box` returns `pred` in `(bx, by, bw, bh)` format. but that's only used to calculate the iou. when it comes to delta, it's still using the raw model output `x` which  is cell offset.

but in your yolo_loss function:
```
        xy_loss = obj_mask * box_loss_scale * \
            tf.reduce_sum(tf.square(true_xy - pred_xy), axis=-1)
        wh_loss = obj_mask * box_loss_scale * \
            tf.reduce_sum(tf.square(true_wh - pred_wh), axis=-1)
```
it just uses `pred_xy` derived from `yolo_boxes ` directly to calculate square error. i wonder what's your consideration in this part?

original yolo code:
```c
float delta_yolo_box(box truth, float *x, float *biases, int n, int index, int i, int j, int lw, int lh, int w, int h, float *delta, float scale, int stride)
{
    box pred = get_yolo_box(x, biases, n, index, i, j, lw, lh, w, h, stride);
    float iou = box_iou(pred, truth);

    float tx = (truth.x*lw - i);
    float ty = (truth.y*lh - j);
    float tw = log(truth.w*w / biases[2*n]);
    float th = log(truth.h*h / biases[2*n + 1]);

    delta[index + 0*stride] = scale * (tx - x[index + 0*stride]);
    delta[index + 1*stride] = scale * (ty - x[index + 1*stride]);
    delta[index + 2*stride] = scale * (tw - x[index + 2*stride]);
    delta[index + 3*stride] = scale * (th - x[index + 3*stride]);
    return iou;
}
```

**Comment by ethanyanjiali** [comment]
i think i get it now, just reviewed the `yolo_boxes` function again, the `pred_box` is actually from the raw model output and get returned directly, so it still makes sense here.
